cnxdb/migrations/20170812000000_current-modules-view.py

```diff
@@ -0,0 +1,86 @@
+# -*- coding: utf-8 -*-
+
+
+def up(cursor):
+    """Replace current_modules with correct logic"""
+    cursor.execute("""
+DROP VIEW current_modules;
+CREATE OR REPLACE VIEW current_modules AS
+WITH latest_idents (module_ident) AS (
+          SELECT module_ident FROM modules m2 join modulestates ms
+                on m2.stateid = ms.stateid
+                WHERE m2.major_version = (
+                    SELECT max(major_version) FROM modules m3
+                        WHERE m2.uuid = m3.uuid
+                )
+                AND
+                (m2.minor_version IS NULL OR
+                 m2.minor_version = (
+                    SELECT max(minor_version) FROM modules m4
+                        WHERE m2.uuid = m4.uuid AND
+                        m2.major_version = m4.major_version
+                    )
+                )
+                AND ms.statename = 'current'
+        )
+SELECT m.* FROM latest_idents li JOIN modules m
+ON m.module_ident = li.module_ident;
+""")
+
+    cursor.execute("""UPDATE modules SET stateid = 1 WHERE stateid IS NULL""")
+
+    cursor.execute("""CREATE TABLE datamigrations.corrected_latest AS
+            SELECT * FROM current_modules cm
+                WHERE NOT EXISTS (
+                    SELECT 1 FROM latest_modules lm
+                        WHERE cm.module_ident = lm.module_ident)""")
+
+    cursor.execute("""CREATE TABLE datamigrations.wrong_latest AS
+            SELECT * FROM latest_modules WHERE uuid IN
+            (SELECT uuid FROM datamigrations.corrected_latest);
+            DELETE FROM latest_modules WHERE uuid IN
+            (SELECT uuid FROM datamigrations.corrected_latest);
+            INSERT INTO latest_modules
```

Comment:
```
karen@xenial-openstax:/var/cnx/venvs/publishing/src/cnx-db$ dbmigrator --db-connection-string='dbname=cnxarchive user=cnxarchive' migrate --version=20170812000000
context undefined, using current directory name "['cnx-db']"
Running migration 20170812000000 current-modules-view
Traceback (most recent call last):
  File "../../bin/dbmigrator", line 11, in <module>
    sys.exit(main())
  File "/var/cnx/venvs/publishing/local/lib/python2.7/site-packages/dbmigrator/cli.py", line 97, in main
    return args['cmmd'](**args)
  File "/var/cnx/venvs/publishing/local/lib/python2.7/site-packages/dbmigrator/utils.py", line 108, in wrapper
    return func(cursor, *args, **kwargs)
  File "/var/cnx/venvs/publishing/local/lib/python2.7/site-packages/dbmigrator/commands/migrate.py", line 31, in cli_command
    migration)
  File "/var/cnx/venvs/publishing/local/lib/python2.7/site-packages/dbmigrator/utils.py", line 178, in compare_schema
    callback(*args, **kwargs)
  File "/var/cnx/venvs/publishing/local/lib/python2.7/site-packages/dbmigrator/utils.py", line 189, in run_migration
    migration.up(cursor)
  File "cnxdb/migrations/20170812000000_current-modules-view.py", line 44, in up
    SELECT * from datamigrations.corrected_latest""")
psycopg2.IntegrityError: duplicate key value violates unique constraint "lastest_modules_short_id_idx"
DETAIL:  Key (short_id(uuid))=(OlLnl474) already exists.

```

Reply:
Doesn't happen for me - I have no content w/ that short_id on my thin data dump, nor on production. Can you repro plz?

Reply:
Yes, I was able to trigger this bug again.  It's a database dump I got from QA last night.

Reply:
Specifically this is the data that causes the error:

```
(195604, 'm64733', '1.10', 'Index', datetime.datetime(2017, 6, 30, 12, 24, 50, 802049, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=-420, name=None)), datetime.datetime(2017, 6, 30, 12, 24, 50, 802049, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=-420, name=None)), 299, 10, ' ', 'cnxbio', '.', 1, None, None, ['OpenStaxCollege'], ['OpenStaxCollege', 'cnxbio'], ['OSCRiceUniversity'], None, 'CompositeModule', '3a52e797-8ef8-58c4-9f17-a70e0b6fe3fa', 10, 115, None, None, 'ccap-biology', None, None)
```

Reply:
This is in the latest_modules:

```
       198889 | m64733   | 1.10    | Index | 2017-07-26 14:33:21.520072-07 | 2017-07-26 14:33:21.520072-07 |      32002 |        10 |         | cnxbio    | .         |       1 |        | en       | {OpenStaxCollege} | {OpenStaxCollege,cnxbio} | {OSCRiceUniversity} |               | CompositeModule | 3a52e797-8ef8-58c4-9f17-a70e0b6fe3fa |            10 |           115 |                  |         | ccap-biology |       |       
```

Reply:
I was thinking, there's no way this could happen since you delete before insert, but there might be duplicates in the corrected_latest table, and this is how it looks like:

```
(Pdb) cursor.execute('SELECT uuid, count(*) FROM datamigrations.corrected_latest GROUP BY uuid')
(Pdb) cursor.fetchall()
[('cb62028b-6697-5af1-94a5-3e3d55a7adb6', 3L), ('ed4264b2-7860-54f8-bf7e-6ed6e455e83d', 3L), ('47f9169d-fe4e-5cde-a2e1-3e48ffbf3c95', 3L), ('a2d22c3d-7bba-5629-bb1a-a2e5eb375e0f', 6L), ('0bc9d78e-0628-58a9-a9f1-82bc387b73e2', 3L), ('14136498-fe87-591d-9ef9-1395b8c61658', 3L), ('a2d45b5b-7313-5e0e-a957-d149f619a219', 2L), ('97505378-4164-5f37-a748-0351e9ece28f', 3L), ('b71b4ffb-538f-5be6-9531-849e865f57dc', 3L), ('edbd2e5a-08a0-5eb4-a191-56923ca33b4a', 3L), ('d1811d98-b1ad-5530-8066-760316793aa5', 3L), ('7169418f-2f89-543d-ab9a-acb9b402fba0', 3L), ('ba4296b3-d2b5-5755-be7c-a763faefcb86', 3L), ('b8bc98e5-52c2-53a5-8685-050cc1dbfd26', 3L), ('58dafade-d928-5b62-9688-6220e01db1ee', 3L), ('cf93a251-a18c-5f9e-951e-9ebba3ab9cf9', 3L), ('fd6dc37b-dd77-5e56-870c-e56da2d551c7', 3L), ('e7c15a94-9f01-5c0e-9f9f-95f69940c349', 3L), ('82fbf9b9-7dc5-5bb7-b525-a51ab5322b16', 2L), ('f42eb909-3c47-5e87-b7c9-8763fcbcf526', 3L), ('45c1c49e-4e14-574b-99e9-39d6e3b014fb', 3L), ('9d4234a5-0d03-5f72-bfd9-c17ed27507ef', 3L), ('3ee954ab-8df9-586f-812e-cc50adf5db20', 1L), ('29aaf329-601b-5969-bcc8-7f93aa339b43', 3L), ('938c3661-7d01-5bb9-9e2d-1eed94fa7d2f', 3L), ('44fa44b6-8f2c-5e02-9b87-6b460c097c20', 3L), ('6197efb6-171f-59e6-8918-24db748e7b05', 3L), ('845207c4-17cf-5c5d-ad81-95ea8f5ab0ab', 3L), ('293c4c8d-c8e3-5269-8bf9-848961d3c384', 2L), ('b0e7fb7d-65cb-509f-9d33-1377307220b1', 3L), ('f851b00a-d3c4-5a04-85b3-03196f1b98ec', 3L), ('76b07248-2325-57dd-9ce2-d239be4aeb26', 3L), ('b1190e25-2526-5ad5-8946-bd3b7ded1ae3', 3L), ('221f9a6a-1fee-545a-824b-7af097aa9293', 3L), ('3a52e797-8ef8-58c4-9f17-a70e0b6fe3fa', 6L), ('e7fa303e-2419-5569-9ea1-a36c6ce2ae92', 3L), ('10ee6e9b-dfb1-58f7-ac62-97b3b24a3e8c', 3L), ('d790bcde-7feb-5669-99be-1af19dfcc72d', 3L), ('ac37c357-f9bc-57b6-81e0-ec3755bb7a2a', 3L), ('e6cb2126-9406-5f82-bb81-284590c0da1b', 3L), ('1b4eed08-fc08-5abe-8769-0b83989328d4', 3L), ('a2e4c48d-3d8b-54fa-8007-201fe503d75e', 3L), ('ecb22ec7-cf20-5ba7-a752-0b0302ea56ae', 3L), ('3556ae5b-c23d-4cb1-aa40-aea4dc9f0e7f', 1L), ('e6aa980a-5228-5ee3-9de8-cdb28cc6d537', 2L), ('1321c689-b10b-53e7-bd64-e78af7295fcf', 3L), ('39f757c9-07a5-5fb4-9f9a-688ee85e4c09', 3L), ('01f41563-1003-519f-8e10-1a13ca5ce67a', 3L), ('eb84b33a-3d30-575b-8e17-5d0f74d4bc59', 3L), ('c6f51d10-9de9-5a2e-9b13-606238293d37', 3L), ('c0ff4538-3cbe-5d60-a768-c63c68fd3afb', 3L), ('590bb0bb-740b-5e43-a3a2-4ab93a3851eb', 3L), ('eebe0ea1-e591-5563-9b41-b0233562ccae', 3L), ('4bccdce5-6b19-5b77-8b1b-8ad80cb960f1', 3L), ('1bf50d94-239f-586d-a470-ba2e897705fe', 3L), ('4e0ca23d-980c-54ff-b92e-94c784ee7846', 3L), ('d5ef636e-a1ab-5638-80ee-7642f5a53868', 3L), ('a2ea60c6-49cd-528f-9f60-5cd2b9b24826', 1L), ('4c6716c9-a292-5c0f-9730-d55308d73816', 3L), ('303e9967-1762-5ea0-8ba0-b00311df4f8c', 2L), ('ffea3b9e-3d7d-57b0-9493-07723e003660', 3L), ('5207095e-cbe7-575b-84d2-85b8ec336f08', 3L), ('8cb09e69-9f55-5260-a35b-198e8f9c46da', 3L), ('dae535f6-871f-562a-b244-f16da036e13f', 3L), ('67c30ef7-90f3-5a60-a084-e10d20a6dcf3', 3L), ('ee01b549-271b-5f4c-af09-cebad8882296', 3L), ('f1c14e75-508c-569f-ae57-c1513aca7446', 3L), ('b48b1720-92cf-5721-bc6e-fd6b70fa03cc', 3L), ('938812df-3c9f-5ccd-a63a-3e8b2145edd8', 3L), ('4d9e0ba6-f72d-5344-9a97-7d5586f15eff', 4L), ('25f02c8c-bba2-5ebb-9548-3333675a9b1b', 3L), ('20df7126-42e5-51b6-acbe-0c67b87f3fff', 2L), ('42296b89-f9e2-588d-b74a-821eff606b3d', 3L), ('af32024b-ec6c-55c4-8693-7bd93628cb91', 3L), ('7cac6e30-83ba-51ce-8017-0f96db12d6d6', 3L), ('969602c0-8a79-54c0-88f5-eadd1b712681', 2L), ('54f8bbd6-fc2f-5605-8bfb-6d7c0c462d55', 3L), ('4b309a3b-2372-5e23-81bd-522f68dbd2ca', 3L), ('8008f854-78f8-5d08-97e7-b6f9cebeb6c9', 3L), ('152e177b-5a1f-5420-bf08-c057266eb5d3', 3L), ('50948675-844c-5353-b9d1-03066497aefa', 3L), ('8d254d53-d27d-5a2c-927e-a55b15552eaa', 3L), ('5266b9d9-1b36-5f88-97dd-935a4a001f39', 3L), ('19aa5cee-852c-5918-8b63-e16bfdb69082', 3L), ('d0c680f7-7376-5b30-a75b-bf0705753c7a', 3L), ('0f201580-d3b9-5094-b09c-0a74c0a4f065', 3L), ('40cc8ff8-e044-51a2-9ce8-da644a529e91', 3L), ('06267b64-adb8-5e03-a5d6-f9b9847c39b4', 3L), ('f01a23d8-3229-5a18-94d4-361e059b63be', 3L), ('09450418-d168-5d35-b497-ef50e4c9897f', 3L), ('8dd1200a-bb33-5de5-ab1a-62416ba623f6', 2L), ('330815a3-d781-5c4f-9064-976452827076', 3L), ('7d9d3083-e90e-5338-b238-a5706e6fc164', 3L), ('437cb858-dbc7-5cd2-8dfe-d73647101538', 3L), ('fe2307e0-a995-53e0-988a-57da627aed5b', 3L), ('e55943db-c318-5da2-a48b-ecab78fca478', 3L), ('3c438835-1508-55ee-b9d7-07db87a700dd', 3L), ('e57a085b-517c-58a6-9ffb-7101112ad4dc', 3L), ('e5053d11-bc5a-5b9c-b1d1-a2e085359873', 3L), ('30a82da1-3c1d-5dbc-88bd-0bf77235b51a', 3L), ('8bf8f1a8-58de-5da4-893b-96590cdc005c', 3L), ('162f41b3-8777-519d-b661-7fb0dc4e2566', 3L), ('24614523-9eec-5bef-a973-5cf443fedd49', 1L), ('5a65396d-b055-5dfe-a616-f6eef549cd72', 3L), ('e6f5703e-aea1-58c4-8245-611ecac58ddc', 3L), ('65e41e48-a368-5fa2-8ef2-10d370b82fd7', 3L), ('8ae5f6d6-7a30-5916-9c87-485c98bc57b6', 3L), ('61b74f94-70ed-5e66-9daa-424148a16ed3', 3L), ('d6518edc-84e8-5e04-a522-0babb9346bb5', 3L), ('72a3ef21-e30b-5ba4-9ea6-eac9699a2f09', 2L), ('d2357f2f-afb9-5370-a607-d615625185ee', 3L), ('72c36a16-868b-5fd5-94ae-8deb00f61b4e', 3L), ('4e4f8db2-ea30-5aff-bb25-085a577510b7', 3L)]
```

Should there be multiple rows for one uuid?

Reply:
I agree, I'll look at this on QA - the migration should be robust to this.

Reply:
BTW: "SELECT uuid, count(*) FROM datamigrations.corrected_latest GROUP BY uuid HAVING count(*) > 1" would be a nice way to see that.

Reply:
This is definitely broken data:

```
(Pdb) cursor.execute('SELECT uuid, array_agg(module_version(major_version, minor_version)), array_agg(module_ident) FROM datamigrations.corrected_latest GROUP BY uuid HAVING 
count(*) > 1')
(Pdb) cursor.fetchall()
[('cb62028b-6697-5af1-94a5-3e3d55a7adb6', ['18.55', '18.55', '18.55'], [195060, 197496, 197594]), ('ed4264b2-7860-54f8-bf7e-6ed6e455e83d', ['18.55', '18.55', '18.55'], [19508
2, 197518, 197616]), ('47f9169d-fe4e-5cde-a2e1-3e48ffbf3c95', ['18.55', '18.55', '18.55'], [195048, 197484, 197582]), ('a2d22c3d-7bba-5629-bb1a-a2e5eb375e0f', ['10.115', '10.
115', '10.115', '10.115', '10.115', '10.115'], [195840, 195603, 195366, 196887, 197162, 198014]), ('0bc9d78e-0628-58a9-a9f1-82bc387b73e2', ['18.55', '18.55', '18.55'], [19507
9, 197515, 197613]), ('14136498-fe87-591d-9ef9-1395b8c61658', ['18.55', '18.55', '18.55'], [195061, 197497, 197595]), ('a2d45b5b-7313-5e0e-a957-d149f619a219', ['11.345', '11.
345'], [196043, 197365]), ('97505378-4164-5f37-a748-0351e9ece28f', ['18.55', '18.55', '18.55'], [195070, 197506, 197604]), ('b71b4ffb-538f-5be6-9531-849e865f57dc', ['18.55', 
'18.55', '18.55'], [195042, 197478, 197576]), ('edbd2e5a-08a0-5eb4-a191-56923ca33b4a', ['18.55', '18.55', '18.55'], [195096, 197532, 197630]), ('d1811d98-b1ad-5530-8066-76031
6793aa5', ['18.55', '18.55', '18.55'], [195083, 197519, 197617]), ('7169418f-2f89-543d-ab9a-acb9b402fba0', ['18.55', '18.55', '18.55'], [195084, 197520, 197618]), ('ba4296b3-
d2b5-5755-be7c-a763faefcb86', ['18.55', '18.55', '18.55'], [195046, 197482, 197580]), ('b8bc98e5-52c2-53a5-8685-050cc1dbfd26', ['18.55', '18.55', '18.55'], [195106, 197542, 1
97640]), ('58dafade-d928-5b62-9688-6220e01db1ee', ['18.55', '18.55', '18.55'], [195066, 197502, 197600]), ('cf93a251-a18c-5f9e-951e-9ebba3ab9cf9', ['18.55', '18.55', '18.55']
, [195053, 197489, 197587]), ('fd6dc37b-dd77-5e56-870c-e56da2d551c7', ['18.55', '18.55', '18.55'], ...
```

Reply:
AS mentioned above, this is solved by removing orphaned CompositeModules. We probably need to make sure that can't happen on production, though.